microbench/src/main/java/io/netty/microbench/channel/AutoFlushBenchmark.java

```diff
@@ -0,0 +1,128 @@
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelOption;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+
+import java.util.concurrent.TimeoutException;
+
+import static java.util.concurrent.TimeUnit.*;
+
+@State(Scope.Benchmark)
+public class AutoFlushBenchmark extends AbstractChannelBenchmark {
+
+    @Param({ "true", "false" })
+    public boolean flush;
+
+    @Param({ "1", "10", "100" })
+    public int writeCount;
+
+    @Setup(Level.Trial)
+    public void setup() {
+        setup0(EMPTY_INITIALIZER, new ChannelInitializer<Channel>() {
+            @Override
+            protected void initChannel(Channel ch) throws Exception {
+                ch.config().setOption(ChannelOption.AUTO_FLUSH, !flush);
+            }
+        });
+    }
+
+    @Benchmark
+    public void compareWithFlushOnEach() throws InterruptedException, TimeoutException {
+        ChannelFuture lastWriteFuture = clientChannel.voidPromise();
+        if (flush) {
+            for (int i = 0; i < writeCount; i++) {
+                lastWriteFuture = pipeline.writeAndFlush(payload.retain());
+            }
+        } else {
+            for (int i = 0; i < writeCount; i++) {
+                lastWriteFuture = pipeline.write(payload.retain());
+            }
+        }
+
+        awaitCompletion(lastWriteFuture);
+    }
+
+    @Benchmark
+    public void compareWithFlushAtEnd() throws InterruptedException, TimeoutException {
+        ChannelFuture lastWriteFuture = clientChannel.voidPromise();
+        if (flush) {
+            for (int i = 0; i < writeCount; i++) {
+                lastWriteFuture = pipeline.write(payload.retain());
+            }
+            pipeline.flush();
+        } else {
+            for (int i = 0; i < writeCount; i++) {
+                lastWriteFuture = pipeline.write(payload.retain());
+            }
+        }
+
+        awaitCompletion(lastWriteFuture);
+    }
+
+    @Benchmark
+    public void compareWithFlushEvery5() throws InterruptedException, TimeoutException {
+        ChannelFuture lastWriteFuture = clientChannel.voidPromise();
+        if (flush) {
+            for (int i = 0; i < writeCount; i++) {
+                lastWriteFuture = pipeline.write(payload.retain());
+                if (i % 5 == 0) {
+                    pipeline.flush();
+                }
+            }
+            pipeline.flush();
+        } else {
+            for (int i = 0; i < writeCount; i++) {
+                lastWriteFuture = pipeline.write(payload.retain());
+            }
+        }
+
+        awaitCompletion(lastWriteFuture);
+    }
+
+    @Benchmark
+    public void compareWithFlushEverySecond() throws InterruptedException, TimeoutException {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this method to throw at most one checked exception instead of: java.lang.InterruptedException, java.util.concurrent.TimeoutException [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1160)